Sprint 5

AlbertKhaydarov opened the pull request:
Сделал отдельный класс QuestionFactory с протоколом и делегатом
Сделал модель ResultAlertModel (QuizResultsViewModel отключил, но не убирал пока)
Сделал отдельный класс ResultAlertPresenter с протоколом и делегатом (в уроке необходимо было выводить промежуточный текст для отработки,  его отработал, потом заменил на целевой по макету)
Сделал модель GameRecordModel
Сделал отдельный класс QuestionFactory 
Сделал отдельный класс QuestionFactory StatisticService с протоколом

Приложение запускается.
Приложение выглядит корректно .
Обработка нажатия на кнопку алерт реализована
Реализована функция сохранения лучшего результата store (с проверкой на то, что новый лучший результат сохранён в UserDefaults). К самому алерту в поле message передан сформированный из рекордной игры текст.
По окончании раунда вызывается алерт, который соответствует макету и на котором отображаются следующие данные:
Текущий результат в виде соотношения правильно отвеченных в раунде вопросов и общего количества вопросов в одном раунде.
Количество сыгранных квизов за всё время, что установлено приложение (при удалении приложения и запуске статистика обнуляется).
Показывает лучший результат за всё время, что приложение установлено (сравнение через ext структуры GameRecord с использованием протокола Comparable). В той же строке выводится дата достижения этого результата в формате, соответствующему макетам.
Средняя точность, рассчитывающаяся как отношение правильно отвеченных вопросов за всё время игры к общему количеству вопросов за всё время игры. Точность выражена в процентах.
Класс StatisticServiceImplementation закрыт протоколом StatisticServiceProtocol, и в MovieQuizViewController используется переменная statisticService типа StatisticServiceProtocol.
При завершении приложение, выкинув его из памяти и запустив снова результаты, которые показываются в конце законченного раунда, не обнуляются.

Commit:
созадана фабрика вопросов

AlbertKhaydarov committed:
доделал  QuestionFactory и добавил QuestionFactoryProtocol

AlbertKhaydarov committed:
делаю делегата для фабрики

Commit:
доделал делегата для фабрики

AlbertKhaydarov committed:
делаю AlertPresenter

Commit:
доделал AlarmPresenter

AlbertKhaydarov committed:
делаю userDefaults

AlbertKhaydarov committed:
делаю UesrDefaults

AlbertKhaydarov committed:
мелочи

AlbertKhaydarov committed:
обновил функцию показа Alert, реализую статистики

AlbertKhaydarov committed:
доделал класс StatisticService

AlbertKhaydarov committed:
мелочи

AlbertKhaydarov committed:
вынес StatisticServiceProtocol в отдельный файл

AlbertKhaydarov committed:
поменял название переменной в MovieQuizViewController на  statisticService

AlbertKhaydarov committed:
добавлено Result(Alert...) в название презентера, протокол и делегат

LizaIlinykh commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,6 +1,8 @@
 
-final class MovieQuizViewController: UIViewController {
+
+final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate, ResultAlertPresenterDelegate {
```
реализацию протоколов лучше выносить в расширения
```
extension MovieQuizViewController: QuestionFactoryDelegate {
// реализация методов протокола QuestionFactoryDelegate
}
```

AlbertKhaydarov replied:
сделал

LizaIlinykh commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -11,96 +13,72 @@ final class MovieQuizViewController: UIViewController {
+            statisticService?.store(correct: correctAnswers, total: questionsAmount)
+            guard let gamesCount = statisticService?.gamesCount else{return}
+            guard let bestgames = statisticService?.bestGame else{return}
+            guard let totalAccuracy = statisticService?.totalAccuracy else{return}
```
else { return } - важно соблюдать код-стайл, отступы в том числе =)

AlbertKhaydarov replied:
сделал

LizaIlinykh commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,61 @@
+//
+
+import Foundation
+class QuestionFactory: QuestionFactoryProtocol {
```
final class

AlbertKhaydarov replied:
сделал

LizaIlinykh commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -0,0 +1,61 @@
+import Foundation
+class QuestionFactory: QuestionFactoryProtocol {
+    
+    weak var delegate: QuestionFactoryDelegate?
```
если вносим через инициализатор, стоит delegate сделать приватным

AlbertKhaydarov replied:
сделал

LizaIlinykh commented on MovieQuiz/Services/StatisticService.swift:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+    
+    var bestGame: GameRecord{
```
предлагаю сделать private(set) чтобы скрыть от записи и записывать только с помощью метода store

AlbertKhaydarov replied:
сделал

AlbertKhaydarov committed:
Исправил баг для первого вопроса в лейбле 10/10 -  currentQuestionIndex = 0 не срабатывал (перенес в  метод  private func showNextQuestionOrResults

реализацию протоколов (делегатов) в MovieQuizViewController вынес в расширения

добавил final для  class QuestionFactory

сделал private(set) var bestGame: GameRecord, чтобы скрыть от записи и записывать только с помощью метода store

поправил отступы в строках 60,61,62 (для скобок после else)

Сделал weak  private var delegate: QuestionFactoryDelegate?
